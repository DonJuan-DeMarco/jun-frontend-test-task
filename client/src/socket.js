import io from 'socket.io-client';

let socket;

const startSocket = () => {
	socket.disconnect();
	socket.connect();
	socket.emit('start');
};

const connectSocket = () => {
	socket = io.connect('http://localhost:3002');
	socket.emit('start');

	socket.on('connect_error', (error) => {
		console.warn('connect_error', error);
	});
	socket.on('disconnect', (reason) => {
		console.warn('connect_error', reason);
	});
	socket.io.on('error', (error) => {
		console.warn('error', error);
	});
	socket.io.on('reconnect_error', (error) => {
		console.warn('reconnect_error', error);
	});
};

export { socket, connectSocket, startSocket };
