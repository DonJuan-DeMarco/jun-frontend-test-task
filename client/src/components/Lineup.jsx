import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { RandomColor } from '../randomColor';
import { socket } from '../socket';
import { horseActions } from '../store/horses';
import ProgressBar from './ProgressBar';

const Lineup = () => {
	const dispatch = useDispatch();
	const horses = useSelector((state) => state.horses.list);
	const order = useSelector((state) => state.horses.finishOrder);

	useEffect(() => {
		if (horses.length === order.length && horses.length !== 0) {
			socket.disconnect();
		}
	}, [horses.length, order.length]);

	return (
		<div className='runningField' data-testid='runningField'>
			{horses.map((horse, i) => {
				if (
					horse.distance === 1000 &&
					(!order.length || order.indexOf(horse.name) < 0)
				) {
					dispatch(horseActions.finish(horse.name));
				}
				return (
					<ProgressBar
						key={i}
						name={horse.name}
						bgcolor={RandomColor(horse.name)}
						distance={horse.distance}
					/>
				);
			})}
		</div>
	);
};

export default Lineup;
