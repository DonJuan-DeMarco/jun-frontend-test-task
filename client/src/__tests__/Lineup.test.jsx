import { configureStore, createSlice } from '@reduxjs/toolkit';
import { render, screen, cleanup } from '@testing-library/react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import Lineup from '../components/Lineup';
import { connectSocket } from '../socket';

afterEach(() => {
	cleanup();
});

const horseSlice = createSlice({
	name: 'horses',
	initialState: {
		list: [
			{
				name: 'Princess Diana',
				distance: 220,
			},
			{
				name: 'Cricket',
				distance: 33,
			},
		],
		finishOrder: ['Princess Diana', 'Cricket'],
	},
	reducers: {
		updateDistance(state, action) {
			state.list = action.payload;
		},
		finish(state, action) {
			state.finishOrder.push(action.payload);
		},
		reset(state) {
			state.finishOrder = [];
			state.list = [];
		},
	},
});

const store = configureStore({
	reducer: {
		horses: horseSlice.reducer,
	},
});
connectSocket();
test('Should render Lineup component', () => {
	render(
		<Provider store={store}>
			<Lineup />
		</Provider>
	);
	const lineupElement = screen.getByTestId('runningField');
	expect(lineupElement).toBeInTheDocument();
	const tree = renderer
		.create(
			<Provider store={store}>
				<Lineup />
			</Provider>
		)
		.toJSON();
	expect(tree).toMatchSnapshot();
	console.log(tree);
});
