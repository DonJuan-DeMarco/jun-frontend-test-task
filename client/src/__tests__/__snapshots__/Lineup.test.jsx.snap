// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render Lineup component 1`] = `
<div
  className="runningField"
  data-testid="runningField"
>
  <div
    className="containerStyles"
    data-testid="progress-bar-Princess Diana"
  >
    <span
      className="labelStyles"
      style={
        Object {
          "color": "#2c34",
        }
      }
    >
      Princess Diana run 220 meters
    </span>
    <div
      className="fillerStyles"
      style={
        Object {
          "backgroundColor": "#2c34",
          "width": "22%",
        }
      }
    />
  </div>
  <div
    className="containerStyles"
    data-testid="progress-bar-Cricket"
  >
    <span
      className="labelStyles"
      style={
        Object {
          "color": "#29d5",
        }
      }
    >
      Cricket run 33 meters
    </span>
    <div
      className="fillerStyles"
      style={
        Object {
          "backgroundColor": "#29d5",
          "width": "3.3%",
        }
      }
    />
  </div>
</div>
`;
